tiledb/sm/metadata/metadata.cc

```diff
@@ -0,0 +1,279 @@
+      offset = sizeof(char);
+      std::memcpy(value_vec.data() + offset, &value_type, sizeof(char));
+      offset += sizeof(char);
+      std::memcpy(value_vec.data() + offset, &value_num, sizeof(value_num));
+      offset += sizeof(value_num);
+      std::memcpy(value_vec.data() + offset, value.data(), value_len);
+
+      // Insert to metadata
+      metadata_map_.emplace(
+          std::make_pair(std::string(key), std::move(value_vec)));
+    }
+  }
+
+  // Create metadata index for fast lookups from index
+  metadata_index_.resize(metadata_map_.size());
+  size_t i = 0;
+  for (auto& m : metadata_map_)
+    metadata_index_[i++] = std::make_pair(&(m.first), &(m.second));
+
+  // Note: `metadata_map_` and `metadata_index_` are immutable after this point
+
+  return Status::Ok();
+}
+
+Status Metadata::serialize(Buffer* buff) const {
+  // Do nothing if there are no metadata to serialize
+  if (metadata_map_.empty())
+    return Status::Ok();
+
+  for (const auto& meta : metadata_map_) {
+    auto key_len = (uint32_t)meta.first.size();
+    RETURN_NOT_OK(buff->write(&key_len, sizeof(uint32_t)));
+    RETURN_NOT_OK(buff->write(meta.first.data(), meta.first.size()));
+    RETURN_NOT_OK(buff->write(meta.second.data(), meta.second.size()));
+  }
+
+  return Status::Ok();
+}
+
+const std::pair<uint64_t, uint64_t>& Metadata::timestamp_range() const {
+  return timestamp_range_;
+}
+
+Status Metadata::del(const char* key) {
+  assert(key != nullptr);
+
+  std::unique_lock<std::mutex> lck(mtx_);
+
+  std::vector<uint8_t> value_vec;
+  value_vec.resize(sizeof(char));
+  char del = 1;
+  std::memcpy(value_vec.data(), &del, sizeof(char));
+  metadata_map_.emplace(std::make_pair(std::string(key), std::move(value_vec)));
+
+  return Status::Ok();
+}
+
+Status Metadata::put(
+    const char* key,
+    Datatype value_type,
+    uint32_t value_num,
+    const void* value) {
+  assert(key != nullptr);
+  assert(value_type != Datatype::ANY);
+  assert(value_num != 0);
+  assert(value != nullptr);
+
+  std::unique_lock<std::mutex> lck(mtx_);
+
+  size_t value_size = value_num * datatype_size(value_type);
+  size_t value_vec_size = 2 * sizeof(char) + sizeof(uint32_t) + value_size;
+  std::vector<uint8_t> value_vec;
+  value_vec.resize(value_vec_size);
+  char del = 0;
+  auto type = static_cast<char>(value_type);
+  std::memcpy(value_vec.data(), &del, sizeof(char));
+  size_t offset = sizeof(char);
+  std::memcpy(value_vec.data() + offset, &type, sizeof(char));
+  offset += sizeof(char);
+  std::memcpy(value_vec.data() + offset, &value_num, sizeof(value_num));
+  offset += sizeof(value_num);
+  std::memcpy(value_vec.data() + offset, value, value_size);
+  metadata_map_.emplace(std::make_pair(std::string(key), std::move(value_vec)));
+
+  return Status::Ok();
+}
+
+Status Metadata::get(
+    const char* key,
+    Datatype* value_type,
+    uint32_t* value_num,
+    const void** value) {
+  assert(key != nullptr);
+
```

Comment:
Do you need to hold the lock in these `get` functions?

Reply:
No because the metadata becomes immutable after it is loaded/deserialized in the READ mode.